pkg/cloud/azure/actuators/machine_scope.go

```diff
@@ -160,26 +161,37 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
 	return m.MachineClient.UpdateStatus(machine)
 }
 
-// Persist the machine spec and machine status.
-func (m *MachineScope) Persist() error {
-	if m.MachineClient == nil {
-		return fmt.Errorf("machine client is empty")
+func (s *MachineScope) PersistIfNeeded(currentMachine *machinev1.Machine) error {
+	currentConfig, err := MachineConfigFromProviderSpec(currentMachine.Spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to get machine config: %v", err)
 	}
-
-	latestMachine, err := m.storeMachineSpec(m.Machine)
+	currentProviderStatus, err := v1beta1.MachineStatusFromProviderStatus(currentMachine.Status.ProviderStatus)
 	if err != nil {
-		return fmt.Errorf("[machinescope] failed to update machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+		return fmt.Errorf("failed to get machine providerStatus: %v", err)
 	}
 
-	_, err = m.storeMachineStatus(latestMachine)
-	if err != nil {
-		return fmt.Errorf("[machinescope] failed to store provider status for machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+	modifiedMachine := s.Machine.DeepCopy()
+	if !equality.Semantic.DeepEqual(currentMachine, modifiedMachine) ||
+		!equality.Semantic.DeepEqual(currentConfig, s.MachineConfig) {
+		modifiedMachine, err = s.storeMachineSpec(modifiedMachine)
```

Comment:
IIUC, the machine object is updated to reflect the new machine spec (assuming `storeMachineSpec` is successful) but we don't refelct that in `s.Machine`. If you now go on to use `s.Machine` (here, or from the callers perspective) then this is now out of date.

Reply:
We should reflect `modifiedMachine` into `s.Machine` (using `DeepCopyInto`). If we do that I think I can live with the rest.